Microsoft.Toolkit.Uwp.UI.Controls.Layout/StaggeredLayout/StaggeredLayout.cs

```diff
@@ -0,0 +1,317 @@
+            {
+                return new Size(availableSize.Width, 0.0);
+            }
+
+            var state = (StaggeredLayoutState)context.LayoutState;
+
+            double availableWidth = availableSize.Width;
+            double availableHeight = availableSize.Height;
+
+            double columnWidth = Math.Min(DesiredColumnWidth, availableWidth);
+            if (columnWidth != state.ColumnWidth)
+            {
+                // The items will need to be remeasured
+                state.Clear();
+            }
+
+            state.ColumnWidth = Math.Min(DesiredColumnWidth, availableWidth);
+            int numColumns = Math.Max(1, (int)Math.Floor(availableWidth / state.ColumnWidth));
+
+            // adjust for column spacing on all columns expect the first
+            double totalWidth = state.ColumnWidth + ((numColumns - 1) * (state.ColumnWidth + ColumnSpacing));
+            if (totalWidth > availableWidth)
+            {
+                numColumns--;
+            }
+            else if (double.IsInfinity(availableWidth))
+            {
+                availableWidth = totalWidth;
+            }
+
+            if (numColumns != state.NumberOfColumns)
+            {
+                // The items will not need to be remeasured, but they will need to go into new columns
+                state.ClearColumns();
+            }
+
+            if (RowSpacing != state.RowSpacing)
+            {
+                // If the RowSpacing changes the height of the rows will be different.
+                // The columns stores the height so we'll want to clear them out to
+                // get the proper height
+                state.ClearColumns();
+                state.RowSpacing = RowSpacing;
+            }
+
+            var columnHeights = new double[numColumns];
+            var itemsPerColumn = new int[numColumns];
+            var deadColumns = new HashSet<int>();
+
+            for (int i = 0; i < context.ItemCount; i++)
+            {
+                var columnIndex = GetColumnIndex(columnHeights);
+
+                UIElement element = null;
+                StaggeredItem item = state.GetItemAt(i);
+                if (item.Height == 0)
+                {
+                    // Item has not been measured yet. Get the element and store the values
+                    element = context.GetOrCreateElementAt(i);
+                    element.Measure(new Size(state.ColumnWidth, availableHeight));
+                    item.Height = element.DesiredSize.Height;
+                }
+
+                double spacing = itemsPerColumn[columnIndex] > 0 ? RowSpacing : 0;
+                item.Top = columnHeights[columnIndex] + spacing;
+                double bottom = item.Top + item.Height;
+                columnHeights[columnIndex] = bottom;
+                itemsPerColumn[columnIndex]++;
+                state.AddItemToColumn(item, columnIndex);
```

Comment:
Would it be better here to use attached properties to remember position/height data vs. copying all the items to our own data structures?

Reply:
What would we attach them to? We can't attach to the elements because they get reused. Attaching to the object of the item could work but that still has overhead. Creating a cache of data speeds up calculation of position/measure, which happens **a lot**